File: Python/Programmers/KAKAO/오픈채팅방.py
# 최종적으로 방을 개설한 사람이 보게 되는 메시지를 출력하는 것이므로 key가 유저아이디, value가 닉네임으로 이루어진 dict 생성
# 이후 최종 변환된 닉네임을 저장

def solution(record):
    answer = []
    user = {} # 유저아이디와 닉네임이 저장될 dict
    for r in record:
        r = r.split(" ") # 공백을 기준으로 split
        if r[0] in ["Enter", "Change"]: # r[0]가 Enter, Change 중 하나라면
            user[r[1]] = r[2] # user에 저장 즉, 제일 마지막에 변하는 닉네임이 최종 닉네임
    
    for r in record:
        r = r.split(" ")
        # 문제 조건에 맞게 출력
        tmp = "{}님이 ".format(user[r[1]])
        if r[0] == "Enter":
            answer.append(tmp + "들어왔습니다.")
        elif r[0] == "Leave":
            answer.append(tmp + "나갔습니다.")

    return answer

File: JavaScript/Programmers/KAKAO/오픈채팅방.js
// 문제 조건에 따라 구현하면 된다.

// 문제에서 요구하는 메시지로 변환해주는 함수
function getMessage(userNickName, action) {
    const command = { Enter: "들어왔습니다.", Leave: "나갔습니다." };
    return `${userNickName}님이 ${command[action]}`;
}

function solution(record) {
    const answer = []; // 결과값을 담을 배열
    const history = {}; // 닉네임의 history를 담을 객체
    record.forEach((current) => { // record 순회
        // action이 Leave일 경우, userNickName 값이 undefined가 되는 것을 유의
        const [action, userId, userNickName] = current.split(" ");
        // action이 Leave가 아니라면 history 객체에 userId에 해당하는 userNickName값 저장
        if (action !== "Leave") history[userId] = userNickName;
        // action이 Change일 경우, 최종 결과값에 포함될 필요 없으므로
        // action이 Change가 아닐 경우, [userId, action] 값 형태로 push
        action !== "Change" && answer.push([userId, action]);
    });
    
    // map 메소드를 통해 answer 배열의 값을 문제에서 요구하는 형태
    return answer.map(([userId, action]) => getMessage(history[userId], action));
}
